v2/internal/reconcilers/arm/azure_generic_arm_reconciler_instance.go

```diff
@@ -248,10 +251,59 @@ func (r *azureDeploymentReconcilerInstance) MonitorDelete(ctx context.Context) (
 	return ctrl.Result{Requeue: true, RequeueAfter: retryAfter}, nil
 }
 
-func (r *azureDeploymentReconcilerInstance) BeginCreateOrUpdateResource(ctx context.Context) (ctrl.Result, error) {
+func (r *azureDeploymentReconcilerInstance) BeginCreateOrUpdateResource(
+	ctx context.Context,
+) (ctrl.Result, error) {
 	if r.Obj.AzureName() == "" {
 		err := errors.New("AzureName was not set. A webhook should default this to .metadata.name if it was omitted. Is the ASO webhook service running?")
-		return ctrl.Result{}, conditions.NewReadyConditionImpactingError(err, conditions.ConditionSeverityError, conditions.ReasonFailed)
+		return ctrl.Result{},
+			conditions.NewReadyConditionImpactingError(
+				err, conditions.ConditionSeverityError, conditions.ReasonFailed)
+	}
+
+	// Create a checker for access to the extension point, if required
+	// (We always get a checker back, even if it does nothing, so we don't need to check for nil)
+	checker, refreshRequired := extensions.CreatePreReconciliationChecker(r.Extension, r.alwaysReconcile)
+	if refreshRequired {
+		// Checker requires our resource to have an up-to-date status
+		r.Log.V(Verbose).Info("Refreshing Status of resource")
+		err := r.updateStatus(ctx)
+		if err != nil && !genericarmclient.IsNotFoundError(err) {
+			// We have an error, and it's not because the resource doesn't exist yet
+			// We assume the error isn't fatal, and we'll try again later
+			return ctrl.Result{},
+				conditions.NewReadyConditionImpactingError(
+					err, conditions.ConditionSeverityWarning, conditions.ReasonFailed)
+		}
+	}
+
+	// Run our pre-reconcilation checker
+	check, err := checker(ctx, r.Obj, r.KubeClient, r.ARMClient, r.Log)
+	if err != nil {
+		// Something went wrong running the check. We assume it's not fatal, and we'll try again later
+		// Make sure we return a ReadyConditionImpactingError so that the Ready condition is updated for the user
+		// Ideally any implementation of the checker should return a ReadyConditionImpactingError, but we can't
+		// guarantee that, so we wrap as required
+		impactingError, ok := conditions.AsReadyConditionImpactingError(err)
+		if !ok {
+			impactingError = conditions.NewReadyConditionImpactingError(
+				err,
+				conditions.ConditionSeverityWarning,
+				conditions.ReasonAzureResourceNotReady)
```

Comment:
is `ReasonAzureResourceNotReady` the right `Reason` here? Given we've already issued the GET on the Azure resource above, the Azure resource must exist here? What is "not ready" about it? The actual error details will be included so I think users will probably figure it out - I'm just not following exactly what this `Reason` means.

Reply:
Ooh I see, in the context of specifically checking something like "Is the resource in a good state for me to send an update" this makes sense.